core/server/worker/src/main/java/alluxio/worker/block/TieredBlockStore.java

```diff
@@ -291,51 +308,180 @@ public void requestSpace(long sessionId, long blockId, long additionalBytes)
       throws BlockDoesNotExistException, WorkerOutOfSpaceException, IOException {
     LOG.debug("requestSpace: sessionId={}, blockId={}, additionalBytes={}", sessionId, blockId,
         additionalBytes);
-    RetryPolicy retryPolicy = new TimeoutRetry(FREE_SPACE_TIMEOUT_MS, EVICTION_INTERVAL_MS);
-    while (retryPolicy.attempt()) {
-      Pair<Boolean, BlockStoreLocation> requestResult =
-          requestSpaceInternal(blockId, additionalBytes);
-      if (requestResult.getFirst()) {
-        return;
+
+    // NOTE: a temp block is only visible to its own writer, unnecessary to acquire
+    // block lock here since no sharing
+    try (LockResource r = new LockResource(mMetadataWriteLock)) {
+      TempBlockMeta tempBlockMeta = mMetaManager.getTempBlockMeta(blockId);
+
+      StorageDirView allocationDir = allocateSpace(sessionId,
+          AllocateOptions.forRequestSpace(additionalBytes, tempBlockMeta.getBlockLocation()));
+      if (allocationDir == null) {
+        throw new WorkerOutOfSpaceException(String.format(
+            "Can't reserve more space for block: %d under session: %d.", blockId, sessionId));
+      }
+
+      if (!allocationDir.toBlockStoreLocation().equals(tempBlockMeta.getBlockLocation())) {
+        // If reached here, allocateSpace() failed to enforce 'forceLocation' flag.
+        throw new RuntimeException("Allocation error: location enforcement failed");
```

Comment:
Is `RuntimeException` the correct exception to throw? Can we provide more information in the message?

Reply:
Right. It's more like `IllegalStateException`. Fixed.